Replaces zip library and small refactor of plugin install

tylersmalley opened the pull request:
Previously our plugin installation used over 5,000 file descriptors when analyzing and extracting the archive. To mitigate this, we are moving to a more supported zip library, Yauzl.

In addition to the replacement of the zip library, this refactors the getPackData method. It was extracting the plugin package.json files, reading them, then finally deleting them. This commit performs this all in memory, removing the need for any file system operations during this step.

These changes reduced the installation time by an average of 35%, from 211.18 seconds to 137.8 seconds. The majority of this time is now spent during optimize.

Fixes #10916

tylersmalley commented:
Fixes https://github.com/elastic/kibana/issues/10916

jbudz commented:
jenkins, test it

jbudz commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
-    if (file.type === SYMBOLIC_LINK) return false;
+export function analyzeArchive(archive) {
+  const plugins = [];
+  const regExp = new RegExp('(kibana[\\\\/][^\\\\/]+)[\\\\/]package\.json', 'i');
```
thoughts on making this name more descriptive?  maybe `packageJsonRegex`

jbudz commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
-      files = files.map((file) => file.replace(/\\/g, '/'));
-      resolve(files);
-    });
+        if (extractPath && fileName.startsWith(extractPath)) {
```
this properly makes the kibana dir as top level, but includes other top level dirs in the kibana directory

```
$ ls plugins/x-pack
LICENSE.txt  NOTICE.txt  elasticsearch  index.js  logstash  node_modules  package.json plugins  server
```

epixa commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
-}
+  return new Promise ((resolve, reject) => {
+    yauzl.open(archive, { lazyEntries: true }, function (err, zipfile) {
+      if (err) {
```
Does `yauzl` document possible `err` objects?  If so, we should include some of the information from the error object in the log message rather than swallowing it, like the message/stack trace.  Things like permission errors during install are common enough that we should provide people with some sort of specific feedback where possible.

epixa commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
+            const chunks = [];
+
+            if (err) {
+              return reject(`unable to read ${entry.fileName} within zip`);
```
Same as above re: swallowing errors.

epixa commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
+  return new Promise((resolve, reject) => {
+    yauzl.open(archive, { lazyEntries: true }, function (err, zipfile) {
+      if (err) {
+        return reject(`unable to read ${archive}`);
```
Same as above

epixa commented on src/cli_plugin/install/zip.js:
```diff
@@ -1,98 +1,120 @@
+          // directory file names end with '/'
+          mkdirp(entry.fileName, function (err) {
+            if (err) {
+              return reject(`unable to create directory: ${entry.fileName}`);
```
Same as above

epixa commented:
There are a lot of instances of us swallowing errors in this and logging a potentially less useful error message in their place.  I started marking each one with a comment, but there are a bunch, so I deleted all my comments except the first one.  It applies throughout the changeset, though.

epixa commented:
This test failure seems to be legit.

tylersmalley commented:
@epixa tests should be passing now. In https://github.com/elastic/kibana/pull/11145/commits/d15383cc736ea6de97c70ccdf8292b83ce6ff2b9 I am re-throwing the error message from Yazul. Here is a [list of possible errors](https://github.com/thejoshwolfe/yauzl/tree/master/test/failure) we could expect. We provide a useful error message to the user and I am now throwing the stack trace as well.

![image](https://cloud.githubusercontent.com/assets/40265/24918825/91775450-1e96-11e7-95bc-15989c968cee.png)

@jbudz addressed your feedback

epixa commented:
There's a conflict on this now

tylersmalley commented:
5.x: https://github.com/elastic/kibana/pull/11166

epixa commented:
LGTM
